api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/TaskTranslatorImpl.java

```diff
@@ -173,52 +173,102 @@ private void setOpenmrsTaskFields(FhirTask openmrsTask, Task fhirTask) {
 		}
 		
 		if (!fhirTask.getInput().isEmpty()) {
-			openmrsTask.setInput(fhirTask.getInput().stream().map(this::translateToInputText).collect(Collectors.toSet()));
+			openmrsTask.setInput(fhirTask.getInput().stream().map(this::translateToInput).collect(Collectors.toSet()));
 		}
 		
 		if (!fhirTask.getOutput().isEmpty()) {
-			openmrsTask.setOutput(
-			    fhirTask.getOutput().stream().map(this::translateToOutputReference).collect(Collectors.toSet()));
+			openmrsTask.setOutput(fhirTask.getOutput().stream().map(this::translateToOutput).collect(Collectors.toSet()));
 		}
 		
 		openmrsTask.setName(FhirConstants.TASK + "/" + fhirTask.getId());
 		
 	}
 	
-	private FhirTaskOutput translateToOutputReference(Task.TaskOutputComponent taskOutputComponent) {
-		FhirReference outputReference = referenceTranslator.toOpenmrsType((Reference) taskOutputComponent.getValue());
-		Concept type = conceptTranslator.toOpenmrsType(taskOutputComponent.getType());
+	private FhirTaskOutput translateToOutput(Task.TaskOutputComponent taskOutputComponent) {
 		FhirTaskOutput output = new FhirTaskOutput();
-		
-		output.setValueReference(outputReference);
+		Concept type = conceptTranslator.toOpenmrsType(taskOutputComponent.getType());
 		output.setType(type);
 		
+		if (taskOutputComponent.getValue() instanceof Reference) {
+			FhirReference ref = referenceTranslator.toOpenmrsType((Reference) taskOutputComponent.getValue());
+			output.setValueReference(ref);
+		}
+		if (taskOutputComponent.getValue() instanceof StringType) {
+			output.setValueText(taskOutputComponent.getValue().toString());
+		}
+		if (taskOutputComponent.getValue() instanceof DecimalType) {
+			output.setValueNumeric(((DecimalType) taskOutputComponent.getValue()).getValueAsNumber().doubleValue());
+		}
+		if (taskOutputComponent.getValue() instanceof DateTimeType) {
+			output.setValueDatetime(((DateTimeType) taskOutputComponent.getValue()).getValue());
+		}
+		
 		output.setName("TaskOutputComponent/" + output.getUuid());
 		
 		return output;
 	}
 	
-	private FhirTaskInput translateToInputText(Task.ParameterComponent parameterComponent) {
-		Concept type = conceptTranslator.toOpenmrsType(parameterComponent.getType());
+	private FhirTaskInput translateToInput(Task.ParameterComponent parameterComponent) {
 		FhirTaskInput input = new FhirTaskInput();
-		
+		Concept type = conceptTranslator.toOpenmrsType(parameterComponent.getType());
 		input.setType(type);
-		input.setValueText(parameterComponent.getValue().toString());
+		
+		if (parameterComponent.getValue() instanceof Reference) {
+			FhirReference ref = referenceTranslator.toOpenmrsType((Reference) parameterComponent.getValue());
```

Comment:
Also, would it make sense to make a separate translator to deal with input and output values/codings and the business logic of it?

Reply:
yeah , moved the translation Logic to separate translators